Optimize dependencies and add a Kotlin library plugin

larmie56 opened the pull request:
**What does this PR do?**
Optimize dependencies and add a Kotlin library plugin

**Description of Task to be completed?**
Use the latest version of 3rd party libraries
Add dagger hilt to the project
Add a Kotlin library plugin

larmie56 committed:
Optimize dependencies

Use latest version of 3rd party libraries and add dagger hilt to the project

larmie56 committed:
Add kotlin library plugin

Ezike commented on buildSrc/build.gradle.kts:
```diff
@@ -22,6 +22,7 @@ gradlePlugin {
 dependencies {
     implementation(Plugin.kotlin)
     implementation(Plugin.androidGradle)
+    implementation(Plugin.daggerHilt)
```
👍🏽

Ezike commented on buildSrc/src/main/kotlin/Extensions.kt:
```diff
@@ -24,6 +24,21 @@ internal val PluginContainer.kotlinKapt: Unit
     }
 
+internal val PluginContainer.kotlinLibrary: Unit
+    get() {
+        apply("kotlin-library")
+    }
+
+internal val PluginContainer.javaLibrary: Unit
+    get() {
+        apply("java-library")
+    }
+
```
this is not the name of the kotlin plugin tho. You need to recheck your implementation as this wouldn't work

Ezike replied:
You can remove the java bit

Ezike commented on buildSrc/src/main/kotlin/Extensions.kt:
```diff
@@ -24,6 +24,21 @@ internal val PluginContainer.kotlinKapt: Unit
+
+internal val PluginContainer.daggerHilt: Unit
+    get() {
+        apply("dagger.hilt.android.plugin")
```
👍🏽

Ezike commented on buildSrc/src/main/kotlin/Library.kt:
```diff
@@ -13,17 +13,21 @@ object Library {
     // Square
     val okhttp: String =
-        "com.squareup.okhttp3:logging-interceptor:${Version.okhttp}"
+        "com.squareup.okhttp3:okhttp:${Version.okhttp}"
```
👏🏽

Ezike reviewed (changes requested):
Nice work. For the kotlin library plugin, you would need to check it out again as the current implementation would fail

Ezike commented on buildSrc/src/main/kotlin/Library.kt:
```diff
@@ -13,17 +13,21 @@ object Library {
     // Google
     const val material: String = "com.google.android.material:material:${Version.material}"
+    const val daggerHiltAndroid: String = "com.google.dagger:hilt-android:${Version.daggerHilt}"
+    const val hiltCompiler: String = "com.google.dagger:hilt-android-compiler:${Version.daggerHilt}"
```
The current hilt compiler doesn't have this name anymore. It's good to pay attention to such things like the current names of some 3rd party deps in order to prevent some nerve racking bugs

larmie56 replied:
Okay thank you very much

Ezike committed:
Fix kotlin library plugin and dagger hilt dependency name.
- Move project extensions to individual files
- Add kotlin and java project extensions
- register kotlin lib plugin in build script

Ezike committed:
- rename plugins

Ezike reviewed (approved):
On to the next 🚀
👏🏽 Well done with this
